File: server/mcs_proto.go
package main

import (
	"encoding/binary"
	"errors"
	"fmt"
	"math"
)

// MCS protocol tags.
const (
	TagHeartbeatPing     = 0
	TagHeartbeatAck      = 1
	TagLoginRequest      = 2
	TagLoginResponse     = 3
	TagClose             = 4
	TagIqStanza          = 7
	TagDataMessageStanza = 8
)

// MCS protocol version.
const mcsVersion = 41

// --- Protobuf helpers (hand-rolled, no codegen) ---

// protoField represents a single protobuf field.
type protoField struct {
	fieldNum uint64
	wireType int // 0=varint, 1=64-bit, 2=length-delimited, 5=32-bit
	varint   uint64
	data     []byte
}

// encodeVarint encodes a varint and returns the bytes.
func encodeVarint(v uint64) []byte {
	buf := make([]byte, binary.MaxVarintLen64)
	n := binary.PutUvarint(buf, v)
	return buf[:n]
}

// encodeTag encodes a protobuf field tag (field number + wire type).
func encodeTag(fieldNum uint64, wireType int) []byte {
	return encodeVarint((fieldNum << 3) | uint64(wireType))
}

// encodeStringField encodes a string/bytes protobuf field.
func encodeStringField(fieldNum uint64, data []byte) []byte {
	var buf []byte
	buf = append(buf, encodeTag(fieldNum, 2)...)
	buf = append(buf, encodeVarint(uint64(len(data)))...)
	buf = append(buf, data...)
	return buf
}

// encodeVarintField encodes a varint protobuf field.
func encodeVarintField(fieldNum uint64, v uint64) []byte {
	var buf []byte
	buf = append(buf, encodeTag(fieldNum, 0)...)
	buf = append(buf, encodeVarint(v)...)
	return buf
}

// decodeProtoFields parses raw protobuf bytes into a list of fields.
func decodeProtoFields(data []byte) ([]protoField, error) {
	var fields []protoField
	offset := 0
	for offset < len(data) {
		tagVal, n := binary.Uvarint(data[offset:])
		if n <= 0 {
			return nil, errors.New("invalid varint for tag")
		}
		offset += n

		fieldNum := tagVal >> 3
		wireType := int(tagVal & 0x07)

		f := protoField{fieldNum: fieldNum, wireType: wireType}

		switch wireType {
		case 0: // varint
			v, n := binary.Uvarint(data[offset:])
			if n <= 0 {
				return nil, errors.New("invalid varint for field value")
			}
			offset += n
			f.varint = v
		case 1: // 64-bit
			if offset+8 > len(data) {
				return nil, errors.New("truncated 64-bit field")
			}
			f.data = data[offset : offset+8]
			offset += 8
		case 2: // length-delimited
			length, n := binary.Uvarint(data[offset:])
			if n <= 0 {
				return nil, errors.New("invalid varint for length")
			}
			offset += n
			if uint64(offset)+length > uint64(len(data)) {
				return nil, fmt.Errorf("length-delimited field overflows: need %d, have %d", length, len(data)-offset)
			}
			f.data = data[offset : offset+int(length)]
			offset += int(length)
		case 5: // 32-bit
			if offset+4 > len(data) {
				return nil, errors.New("truncated 32-bit field")
			}
			f.data = data[offset : offset+4]
			offset += 4
		default:
			return nil, fmt.Errorf("unsupported wire type %d", wireType)
		}

		fields = append(fields, f)
	}
	return fields, nil
}

// getStringField returns the first length-delimited field with the given number.
func getStringField(fields []protoField, num uint64) string {
	for _, f := range fields {
		if f.fieldNum == num && f.wireType == 2 {
			return string(f.data)
		}
	}
	return ""
}

// getVarintField returns the first varint field with the given number.
func getVarintField(fields []protoField, num uint64) (uint64, bool) {
	for _, f := range fields {
		if f.fieldNum == num && f.wireType == 0 {
			return f.varint, true
		}
	}
	return 0, false
}

// getRepeatedMessage returns all length-delimited fields with the given number,
// each parsed as a sub-message.
func getRepeatedMessage(fields []protoField, num uint64) [][]protoField {
	var result [][]protoField
	for _, f := range fields {
		if f.fieldNum == num && f.wireType == 2 {
			sub, err := decodeProtoFields(f.data)
			if err == nil {
				result = append(result, sub)
			}
		}
	}
	return result
}

// --- MCS Messages ---

// LoginRequest builds a serialised MCS LoginRequest protobuf.
//
// LoginRequest proto fields (from Android MCS proto):
//
//	field 1 (string):          id — Chrome build identifier
//	field 2 (string):          domain — "mcs.android.com"
//	field 3 (string):          user — android_id as string
//	field 4 (string):          resource — android_id as string
//	field 5 (string):          auth_token — security_token as string
//	field 6 (string):          device_id — "android-HEX"
//	field 7 (int64):           last_rmq_id
//	field 8 (repeated Setting): setting — key-value pairs
//	field 9 (int32):           compress
//	field 10 (repeated string): received_persistent_id
//	field 12 (bool):           adaptive_heartbeat
//	field 14 (bool):           use_rmq2 — CRITICAL: enables message delivery
//	field 15 (int64):          account_id
//	field 16 (int32):          auth_service — 2 = ANDROID_ID
//	field 17 (int32):          network_type — 1 = WiFi
func BuildLoginRequest(androidID, securityToken uint64) []byte {
	hexID := fmt.Sprintf("%x", androidID)
	clientID := "chrome-63.0.3234.0"
	deviceID := "android-" + hexID
	aidStr := fmt.Sprintf("%d", androidID)
	tokenStr := fmt.Sprintf("%d", securityToken)

	// Build Setting message for "new_vc" = "1"
	var settingMsg []byte
	settingMsg = append(settingMsg, encodeStringField(1, []byte("new_vc"))...)
	settingMsg = append(settingMsg, encodeStringField(2, []byte("1"))...)

	var msg []byte
	msg = append(msg, encodeStringField(1, []byte(clientID))...)          // id
	msg = append(msg, encodeStringField(2, []byte("mcs.android.com"))...) // domain
	msg = append(msg, encodeStringField(3, []byte(aidStr))...)            // user
	msg = append(msg, encodeStringField(4, []byte(aidStr))...)            // resource
	msg = append(msg, encodeStringField(5, []byte(tokenStr))...)          // auth_token
	msg = append(msg, encodeStringField(6, []byte(deviceID))...)          // device_id
	msg = append(msg, encodeStringField(8, settingMsg)...)                // setting: new_vc=1
	msg = append(msg, encodeVarintField(14, 1)...)                        // use_rmq2 = true
	msg = append(msg, encodeVarintField(16, 2)...)                        // auth_service = ANDROID_ID
	msg = append(msg, encodeVarintField(17, 1)...)                        // network_type = WiFi

	return msg
}

// HeartbeatPing fields:
//
//	field 1 (int32): stream_id — our outgoing stream counter
//	field 2 (int32): last_stream_id_received — last server message we received
//	field 3 (int64): status
func BuildHeartbeatPing(streamID, lastStreamIDReceived int) []byte {
	var msg []byte
	if streamID > 0 {
		msg = append(msg, encodeVarintField(1, uint64(streamID))...)
	}
	if lastStreamIDReceived > 0 {
		msg = append(msg, encodeVarintField(2, uint64(lastStreamIDReceived))...)
	}
	msg = append(msg, encodeVarintField(3, 0)...) // status = 0
	return msg
}

// HeartbeatAck is sent in response to HeartbeatPing.
func BuildHeartbeatAck(lastStreamIDReceived int) []byte {
	var msg []byte
	if lastStreamIDReceived > 0 {
		msg = append(msg, encodeVarintField(2, uint64(lastStreamIDReceived))...)
	}
	msg = append(msg, encodeVarintField(3, 0)...) // status = 0
	return msg
}

// SelectiveAck acknowledges persistent IDs.
//
// IqStanza fields for selective ack:
//
//	field 2 (int32): type = 1 (SET)
//	field 3 (string): id
//	field 7 (message): extension (id=12, data=<SelectiveAck>)
func BuildSelectiveAck(persistentIDs []string, iqID string) []byte {
	if iqID == "" {
		iqID = "0"
	}

	// Build inner SelectiveAck message: field 1 (repeated string): id
	var innerMsg []byte
	for _, id := range persistentIDs {
		innerMsg = append(innerMsg, encodeStringField(1, []byte(id))...)
	}

	// Build extension message: field 1 (id=12), field 2 (bytes=data)
	var extMsg []byte
	extMsg = append(extMsg, encodeVarintField(1, 12)...)
	extMsg = append(extMsg, encodeStringField(2, innerMsg)...)

	// Build IqStanza
	var msg []byte
	msg = append(msg, encodeVarintField(2, 1)...)            // type = SET
	msg = append(msg, encodeStringField(3, []byte(iqID))...) // id
	msg = append(msg, encodeStringField(7, extMsg)...)       // extension
	return msg
}

// BuildIqResult builds an IqStanza response with type RESULT.
func BuildIqResult(iqID, to, from string) []byte {
	var msg []byte
	msg = append(msg, encodeVarintField(2, 2)...)            // type = RESULT
	msg = append(msg, encodeStringField(3, []byte(iqID))...) // id
	if from != "" {
		msg = append(msg, encodeStringField(4, []byte(from))...)
	}
	if to != "" {
		msg = append(msg, encodeStringField(5, []byte(to))...)
	}
	return msg
}

// AppData is a key-value pair from a DataMessageStanza.
type AppData struct {
	Key   string
	Value string
}

// DataMessage is a parsed DataMessageStanza.
type DataMessage struct {
	From         string
	Category     string
	AppDataList  []AppData
	PersistentID string
}

// ParseDataMessageStanza extracts fields from a DataMessageStanza protobuf.
//
// DataMessageStanza fields:
//
//	field 3 (string): from
//	field 5 (string): category
//	field 7 (repeated message): app_data — each has field 1 (key), field 2 (value)
//	field 9 (string): persistent_id
func ParseDataMessageStanza(data []byte) (*DataMessage, error) {
	fields, err := decodeProtoFields(data)
	if err != nil {
		return nil, err
	}

	msg := &DataMessage{
		From:         getStringField(fields, 3),
		Category:     getStringField(fields, 5),
		PersistentID: getStringField(fields, 9),
	}

	// Parse repeated app_data (field 7).
	for _, sub := range getRepeatedMessage(fields, 7) {
		kv := AppData{
			Key:   getStringField(sub, 1),
			Value: getStringField(sub, 2),
		}
		msg.AppDataList = append(msg.AppDataList, kv)
	}

	return msg, nil
}

// GetAppDataValue returns the value for a given key from a DataMessage.
func (d *DataMessage) GetAppDataValue(key string) string {
	for _, kv := range d.AppDataList {
		if kv.Key == key {
			return kv.Value
		}
	}
	return ""
}

// --- MCS wire format ---

// EncodeMCSMessage wraps a protobuf message with MCS wire framing:
// [1 byte: tag] [varint: length] [message bytes]
// On first message, prepend version byte.
func EncodeMCSMessage(tag byte, msg []byte, includeVersion bool) []byte {
	var buf []byte
	if includeVersion {
		buf = append(buf, mcsVersion)
	}
	buf = append(buf, tag)
	buf = append(buf, encodeVarint(uint64(len(msg)))...)
	buf = append(buf, msg...)
	return buf
}

// MCSReader reads MCS messages from a byte stream.
type MCSReader struct {
	buf         []byte
	versionRead bool
}

// NewMCSReader creates a new reader.
func NewMCSReader() *MCSReader {
	return &MCSReader{}
}

// Feed adds data to the reader's buffer.
func (r *MCSReader) Feed(data []byte) {
	r.buf = append(r.buf, data...)
}

// MCSMessage is a parsed MCS message (tag + body).
type MCSMessage struct {
	Tag  byte
	Body []byte
}

// Next tries to parse the next complete message from the buffer.
// Returns nil if not enough data yet.
func (r *MCSReader) Next() *MCSMessage {
	// Strip version byte from buffer on first read.
	if !r.versionRead {
		if len(r.buf) == 0 {
			return nil
		}
		r.buf = r.buf[1:] // consume version byte from buffer
		r.versionRead = true
	}

	if len(r.buf) < 2 {
		// Need at least tag + 1 byte of varint.
		return nil
	}

	// Read tag.
	tag := r.buf[0]

	// Read varint length.
	length, n := binary.Uvarint(r.buf[1:])
	if n <= 0 {
		return nil
	}

	headerLen := 1 + n // tag byte + varint bytes
	msgLen := int(length)

	if length > math.MaxInt32 {
		r.buf = r.buf[headerLen:]
		return nil
	}

	if headerLen+msgLen > len(r.buf) {
		// Not enough data for the full message body.
		return nil
	}

	body := make([]byte, msgLen)
	copy(body, r.buf[headerLen:headerLen+msgLen])
	r.buf = r.buf[headerLen+msgLen:]

	return &MCSMessage{Tag: tag, Body: body}
}

File: client/lib/mcs_proto.dart
import 'dart:typed_data';

/// MCS protocol tags.
class MCSTag {
  static const int heartbeatPing = 0;
  static const int heartbeatAck = 1;
  static const int loginRequest = 2;
  static const int loginResponse = 3;
  static const int close = 4;
  static const int iqStanza = 7;
  static const int dataMessageStanza = 8;
}

/// MCS protocol version.
const int mcsVersion = 41;

// --- Protobuf helpers (hand-rolled, no codegen) ---

/// A parsed protobuf field.
class ProtoField {
  final int fieldNum;
  final int wireType; // 0=varint, 1=64-bit, 2=length-delimited, 5=32-bit
  final int varint;
  final Uint8List data;

  ProtoField({
    required this.fieldNum,
    required this.wireType,
    this.varint = 0,
    Uint8List? data,
  }) : data = data ?? Uint8List(0);
}

/// Encode a varint.
Uint8List encodeVarint(int v) {
  final bytes = <int>[];
  var value = v;
  while (value > 0x7f) {
    bytes.add((value & 0x7f) | 0x80);
    value >>= 7;
  }
  bytes.add(value & 0x7f);
  return Uint8List.fromList(bytes);
}

/// Encode a protobuf tag.
Uint8List encodeTag(int fieldNum, int wireType) {
  return encodeVarint((fieldNum << 3) | wireType);
}

/// Encode a string/bytes protobuf field.
Uint8List encodeStringField(int fieldNum, Uint8List data) {
  final buf = BytesBuilder();
  buf.add(encodeTag(fieldNum, 2));
  buf.add(encodeVarint(data.length));
  buf.add(data);
  return buf.toBytes();
}

/// Encode a varint protobuf field.
Uint8List encodeVarintField(int fieldNum, int value) {
  final buf = BytesBuilder();
  buf.add(encodeTag(fieldNum, 0));
  buf.add(encodeVarint(value));
  return buf.toBytes();
}

/// Decode a varint from data at offset. Returns (value, bytesConsumed).
(int, int) decodeVarInt(Uint8List data, int offset) {
  int result = 0;
  int shift = 0;
  int i = offset;
  while (i < data.length) {
    final b = data[i];
    result |= (b & 0x7f) << shift;
    i++;
    if ((b & 0x80) == 0) {
      return (result, i - offset);
    }
    shift += 7;
    if (shift >= 64) break;
  }
  throw FormatException('Invalid varint');
}

/// Parse raw protobuf bytes into a list of fields.
List<ProtoField> decodeProtoFields(Uint8List data) {
  final fields = <ProtoField>[];
  int offset = 0;
  while (offset < data.length) {
    final (tagVal, tn) = decodeVarInt(data, offset);
    offset += tn;

    final fieldNum = tagVal >> 3;
    final wireType = tagVal & 0x07;

    switch (wireType) {
      case 0: // varint
        final (v, vn) = decodeVarInt(data, offset);
        offset += vn;
        fields.add(ProtoField(fieldNum: fieldNum, wireType: wireType, varint: v));
      case 1: // 64-bit
        if (offset + 8 > data.length) throw FormatException('Truncated 64-bit');
        fields.add(ProtoField(
          fieldNum: fieldNum,
          wireType: wireType,
          data: Uint8List.fromList(data.sublist(offset, offset + 8)),
        ));
        offset += 8;
      case 2: // length-delimited
        final (length, ln) = decodeVarInt(data, offset);
        offset += ln;
        if (offset + length > data.length) {
          throw FormatException('Length-delimited overflows');
        }
        fields.add(ProtoField(
          fieldNum: fieldNum,
          wireType: wireType,
          data: Uint8List.fromList(data.sublist(offset, offset + length)),
        ));
        offset += length;
      case 5: // 32-bit
        if (offset + 4 > data.length) throw FormatException('Truncated 32-bit');
        fields.add(ProtoField(
          fieldNum: fieldNum,
          wireType: wireType,
          data: Uint8List.fromList(data.sublist(offset, offset + 4)),
        ));
        offset += 4;
      default:
        throw FormatException('Unsupported wire type $wireType');
    }
  }
  return fields;
}

/// Get the first string field with the given number.
String getStringField(List<ProtoField> fields, int num) {
  for (final f in fields) {
    if (f.fieldNum == num && f.wireType == 2) {
      return String.fromCharCodes(f.data);
    }
  }
  return '';
}

/// Get the first varint field with the given number.
int? getVarintField(List<ProtoField> fields, int num) {
  for (final f in fields) {
    if (f.fieldNum == num && f.wireType == 0) {
      return f.varint;
    }
  }
  return null;
}

/// Get all repeated sub-messages for a field number.
List<List<ProtoField>> getRepeatedMessage(List<ProtoField> fields, int num) {
  final result = <List<ProtoField>>[];
  for (final f in fields) {
    if (f.fieldNum == num && f.wireType == 2) {
      try {
        result.add(decodeProtoFields(f.data));
      } catch (_) {}
    }
  }
  return result;
}

// --- MCS Messages ---

/// Build a serialised LoginRequest protobuf.
///
/// Field numbers (from Android MCS proto):
///   1: id, 2: domain, 3: user, 4: resource, 5: auth_token,
///   6: device_id, 8: setting (repeated), 14: use_rmq2,
///   16: auth_service, 17: network_type
Uint8List buildLoginRequest(int androidId, int securityToken) {
  final hexId = androidId.toRadixString(16);
  final clientId = 'chrome-63.0.3234.0';
  final deviceId = 'android-$hexId';
  final aidStr = androidId.toString();
  final tokenStr = securityToken.toString();

  // Build Setting sub-message: name="new_vc", value="1"
  final settingBuf = BytesBuilder();
  settingBuf.add(encodeStringField(1, Uint8List.fromList('new_vc'.codeUnits)));
  settingBuf.add(encodeStringField(2, Uint8List.fromList('1'.codeUnits)));

  final buf = BytesBuilder();
  buf.add(encodeStringField(1, Uint8List.fromList(clientId.codeUnits)));   // id
  buf.add(encodeStringField(2, Uint8List.fromList('mcs.android.com'.codeUnits))); // domain
  buf.add(encodeStringField(3, Uint8List.fromList(aidStr.codeUnits)));    // user
  buf.add(encodeStringField(4, Uint8List.fromList(aidStr.codeUnits)));    // resource
  buf.add(encodeStringField(5, Uint8List.fromList(tokenStr.codeUnits))); // auth_token
  buf.add(encodeStringField(6, Uint8List.fromList(deviceId.codeUnits)));  // device_id
  buf.add(encodeStringField(8, settingBuf.toBytes()));                    // setting: new_vc=1
  buf.add(encodeVarintField(14, 1));  // use_rmq2 = true
  buf.add(encodeVarintField(16, 2));  // auth_service = ANDROID_ID
  buf.add(encodeVarintField(17, 1));  // network_type = WiFi
  return buf.toBytes();
}

/// Build a HeartbeatPing with stream ID tracking.
Uint8List buildHeartbeatPing(int streamId, int lastStreamIdReceived) {
  final buf = BytesBuilder();
  if (streamId > 0) {
    buf.add(encodeVarintField(1, streamId));
  }
  if (lastStreamIdReceived > 0) {
    buf.add(encodeVarintField(2, lastStreamIdReceived));
  }
  buf.add(encodeVarintField(3, 0)); // status = 0
  return buf.toBytes();
}

/// Build a HeartbeatAck with stream ID tracking.
Uint8List buildHeartbeatAck(int lastStreamIdReceived) {
  final buf = BytesBuilder();
  if (lastStreamIdReceived > 0) {
    buf.add(encodeVarintField(2, lastStreamIdReceived));
  }
  buf.add(encodeVarintField(3, 0)); // status = 0
  return buf.toBytes();
}

/// Build a SelectiveAck IqStanza.
Uint8List buildSelectiveAck(List<String> persistentIds, {String iqId = '0'}) {
  if (iqId.isEmpty) {
    iqId = '0';
  }

  // Inner SelectiveAck: field 1 (repeated string): id
  final innerBuf = BytesBuilder();
  for (final id in persistentIds) {
    innerBuf.add(encodeStringField(1, Uint8List.fromList(id.codeUnits)));
  }

  // Extension: field 1=id (12), field 2=data (SelectiveAck bytes)
  final extBuf = BytesBuilder();
  extBuf.add(encodeVarintField(1, 12));
  extBuf.add(encodeStringField(2, innerBuf.toBytes()));

  // IqStanza
  final buf = BytesBuilder();
  buf.add(encodeVarintField(2, 1)); // type = SET
  buf.add(encodeStringField(3, Uint8List.fromList(iqId.codeUnits))); // id
  buf.add(encodeStringField(7, extBuf.toBytes())); // extension
  return buf.toBytes();
}

/// Build an IqStanza RESULT response.
Uint8List buildIqResult(String iqId, {String to = '', String from = ''}) {
  final buf = BytesBuilder();
  buf.add(encodeVarintField(2, 2)); // type = RESULT
  buf.add(encodeStringField(3, Uint8List.fromList(iqId.codeUnits))); // id
  if (from.isNotEmpty) {
    buf.add(encodeStringField(4, Uint8List.fromList(from.codeUnits)));
  }
  if (to.isNotEmpty) {
    buf.add(encodeStringField(5, Uint8List.fromList(to.codeUnits)));
  }
  return buf.toBytes();
}

/// A key-value pair from a DataMessageStanza.
class AppData {
  final String key;
  final String value;
  AppData({required this.key, required this.value});
}

/// A parsed DataMessageStanza.
class DataMessage {
  final String from;
  final String category;
  final List<AppData> appDataList;
  final String persistentId;

  DataMessage({
    required this.from,
    required this.category,
    required this.appDataList,
    required this.persistentId,
  });

  String getAppDataValue(String key) {
    for (final kv in appDataList) {
      if (kv.key == key) return kv.value;
    }
    return '';
  }
}

/// Parse a DataMessageStanza protobuf.
DataMessage parseDataMessageStanza(Uint8List data) {
  final fields = decodeProtoFields(data);

  final appData = <AppData>[];
  for (final sub in getRepeatedMessage(fields, 7)) {
    appData.add(AppData(
      key: getStringField(sub, 1),
      value: getStringField(sub, 2),
    ));
  }

  return DataMessage(
    from: getStringField(fields, 3),
    category: getStringField(fields, 5),
    appDataList: appData,
    persistentId: getStringField(fields, 9),
  );
}

// --- MCS wire format ---

/// Encode an MCS message with wire framing.
/// [tag] [varint length] [message bytes]
/// If [includeVersion], prepend version byte (first message only).
Uint8List encodeMCSMessage(int tag, Uint8List msg, {bool includeVersion = false}) {
  final buf = BytesBuilder();
  if (includeVersion) {
    buf.addByte(mcsVersion);
  }
  buf.addByte(tag);
  buf.add(encodeVarint(msg.length));
  buf.add(msg);
  return buf.toBytes();
}

/// A parsed MCS message (tag + body).
class MCSMessage {
  final int tag;
  final Uint8List body;
  MCSMessage({required this.tag, required this.body});
}

/// Reads MCS messages from a byte stream.
class MCSReader {
  final BytesBuilder _buf = BytesBuilder();
  bool _versionRead = false;

  /// Feed data into the reader's buffer.
  void feed(Uint8List data) {
    _buf.add(data);
  }

  /// Try to parse the next complete message. Returns null if not enough data.
  MCSMessage? next() {
    var data = _buf.toBytes();
    if (data.isEmpty) return null;

    // Strip version byte from buffer on first read.
    if (!_versionRead) {
      _buf.clear();
      if (data.length > 1) {
        _buf.add(data.sublist(1));
      }
      _versionRead = true;
      data = _buf.toBytes();
      if (data.isEmpty) return null;
    }

    if (data.length < 2) return null; // need tag + varint

    // Read tag.
    final tag = data[0];

    // Read varint length.
    int length;
    int n;
    try {
      final result = decodeVarInt(Uint8List.sublistView(data, 1), 0);
      length = result.$1;
      n = result.$2;
    } catch (_) {
      return null;
    }

    final headerLen = 1 + n; // tag + varint
    if (headerLen + length > data.length) return null;

    final body = Uint8List.fromList(data.sublist(headerLen, headerLen + length));

    // Consume processed bytes.
    _buf.clear();
    if (headerLen + length < data.length) {
      _buf.add(data.sublist(headerLen + length));
    }

    return MCSMessage(tag: tag, body: body);
  }
}
